pep-0635.rst

```diff
@@ -0,0 +1,588 @@
+but also on the value of some class attributes, like the ``BinOp``
+example above.  The Visitor pattern is insufficiently flexible for
+this: it can only select based on the class.
+
+For a complete example, see
+https://github.com/gvanrossum/patma/blob/master/examples/expr.py#L231
+
+TODO: Could we say more here?
+
+Pattern and functional style
+----------------------------
+
+Most Python applications and libraries are not written in a consistent
+OO style -- unlike Java, Python encourages defining functions at the
+top-level of a module, and for simple data structures, tuples (or
+named tuples or lists) and dictionaries are often used exclusively or
+mixed with classes or data classes.
+
+Pattern matching is particularly suitable for picking apart such data
+structures.  As an extreme example, it's easy to write code that picks
+a JSON data structure using ``match``.
+
+TODO: Example code.
+
+
+
+Motivation (Tobias' version)
+============================
+
+**This section should explain why we think pattern matching is a good
+addition for Python.**
+
+Since Python is a dynamically typed language, Python code frequently
+has to deal with data that comes in different forms and shapes.  This,
+in turn, gives rise to a high degree of versatility by favouring a
+'duck typing' style, imposing only the bare minimum of requirements
+on the form and shape of data, i.e. the structure of data objects.
+Nonetheless, we find that Python code is often sparkled with
+conditions depending on ``isinstance``, ``len``, ``getattr``,
+``hasattr``, etc.  Despite the benefits of 'duck typing', actual
+code still requires to query the format or type of an object in order
+to choose an appropriate processing action or to extract relevant bits
+of information.
+
+Unfortunately, the conditions for slightly more complex structures
+quickly grow into sequences of partly interdependent bits of
+structural tests, hurting readbility and maintability.  Pattern
+matching offers here a more direct scheme of expressing the minimal
+structure that an objects needs to have in order to allow for further
+processing.  Rather than writing a series of manual tests, patterns
+follow a *declarative* style, which is well-known for improved
+readability, maintability, and for delegating the burden of efficient
+execution on the compiler and interpreter.
+
+The concept of pattern matching is similar to regular expressions,
+where succinct patterns describe a textual structure.  A dedicated
+compiler then transforms these declarative patterns into highly
+efficient finite state machines.  In contrast to regular expressions,
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
+match.  Moreover, pattern matching blends the matching of a suitable
+pattern with that of a function, i.e. code that is executed in order
+to handle and process the information provided by a specific kind of
+object.
```

Comment:
Conceptually, it is a function in that it binds parameters and then executes some code.  Technically, it is not because it does not have its dedicated scope.  But granted: this approach might be too abstract for most readers.

This is similar to for-loops, which conceptually also have an embedded "function" in that they bind parameters and then execute some code (the loop's body).  In _Scala_, for instance, both the body of a loop as well as a case clause are a scope of their own, so that the similarity to a function is very strong.  Other languages (e.g. Grace) have even explicitly defined each case clause to be a function of its own.  Nonetheless, it might feel alien to a Python programmer to see it this way, because there is no dedicated scope in Python and the iteration/pattern variables leak into the surrounding context.

Anyway, my intention behind all this is to really stress the idea of pattern matching as "_extract some data from an object, bind it to new local variables (parameters), and then do something with it_".  In particular as opposed to "choose some action based on the type of an object", because the latter leads to switch semantics.

Reply:
I understand, yes.  I think there is actually even a passage in our paper where we discuss the issue that scope and functions are strongly intertwined in Python.